src/jltypes.c

```diff
@@ -881,29 +881,57 @@ static jl_value_t *inst_datatype(jl_datatype_t *dt, jl_svec_t *p, jl_value_t **i
     if (istuple && ntp > 0 && jl_is_vararg_type(last)) {
         // normalize Tuple{..., Vararg{Int, 3}} to Tuple{..., Int, Int, Int}
         jl_value_t *va = jl_unwrap_unionall(last);
+        jl_value_t *va0 = jl_tparam0(va), *va1 = jl_tparam1(va);
         // return same `Tuple` object for types equal to it
-        if (ntp == 1 && jl_tparam0(va) == (jl_value_t*)jl_any_type &&
-            jl_is_unionall(last) && jl_tparam1(va) == (jl_value_t*)((jl_unionall_t*)last)->var) {
+        if (ntp == 1 &&
+            (last == (jl_value_t*)jl_vararg_type ||  // Tuple{Vararg} == Tuple
+             (va0 == (jl_value_t*)jl_any_type &&
+              jl_is_unionall(last) && va1 == (jl_value_t*)((jl_unionall_t*)last)->var))) {
             if (cacheable) JL_UNLOCK(&typecache_lock); // Might GC
             return (jl_value_t*)jl_anytuple_type;
         }
-        if (jl_is_long(jl_tparam1(va))) {
-            ssize_t nt = jl_unbox_long(jl_tparam1(va));
+        // normalize Tuple{..., Vararg{T,_} where T<:S} to Tuple{..., Vararg{S,_}}
+        if (jl_is_unionall(last)) {
+            jl_tvar_t *tv = ((jl_unionall_t*)last)->var;
+            jl_value_t *vb = ((jl_unionall_t*)last)->body;
+            if (va0 == (jl_value_t*)tv) {
+                va0 = tv->ub;
+                // TODO: maybe shouldn't overwrite this
+                last = iparams[ntp - 1] = jl_wrap_vararg(va0, va1);
```

Comment:
Conveniently, `jl_f_tuple` only makes concrete types :) But yes, I'll change this.